addons/purchase/tests/test_purchase_downpayment.py

```diff
@@ -0,0 +1,686 @@
+            'product_uom': cls.product_order.uom_id.id,
+            'price_unit': 100,
+            'order_id': cls.purchase_order.id,
+            'tax_ids': False,
+        })
+        cls.pol_serv_deliver = cls.env['purchase.order.line'].create({
+            'name': cls.service_deliver.name,
+            'product_id': cls.service_deliver.id,
+            'product_qty': 2,
+            'product_uom': cls.service_deliver.uom_id.id,
+            'price_unit': 100,
+            'order_id': cls.purchase_order.id,
+            'tax_ids': False,
+        })
+        cls.pol_serv_order = cls.env['purchase.order.line'].create({
+            'name': cls.service_order.name,
+            'product_id': cls.service_order.id,
+            'product_qty': 2,
+            'product_uom': cls.service_order.uom_id.id,
+            'price_unit': 100,
+            'order_id': cls.purchase_order.id,
+            'tax_ids': False,
+        })
+        cls.pol_product_deliver = cls.env['purchase.order.line'].create({
+            'name': cls.product_deliver.name,
+            'product_id': cls.product_deliver.id,
+            'product_qty': 2,
+            'product_uom': cls.product_deliver.uom_id.id,
+            'price_unit': 100,
+            'order_id': cls.purchase_order.id,
+            'tax_ids': False,
+        })
+
+        cls.expense_account = cls.company_data['default_account_expense']
+        cls.payable_account = cls.company_data['default_account_payable']
+
+        cls.context = {
+            'active_model': 'purchase.order',
+            'active_ids': [cls.purchase_order.id],
+            'active_id': cls.purchase_order.id,
+            'default_journal_id': cls.company_data['default_journal_purchase'].id,
+        }
+
+    @classmethod
+    def create_tax(cls, amount, values=None):
+        vals = {
+            'name': f'Tax {amount} {uuid.uuid4()}',
+            'amount_type': 'percent',
+            'amount': amount,
+            'type_tax_use': 'purchase',
+            'repartition_line_ids': [
+                Command.create({'document_type': 'invoice', 'repartition_type': 'base'}),
+                Command.create({'document_type': 'invoice', 'repartition_type': 'tax', 'account_id': cls.tax_account.id}),
+                Command.create({'document_type': 'refund', 'repartition_type': 'base'}),
+                Command.create({'document_type': 'refund', 'repartition_type': 'tax', 'account_id': cls.tax_account.id}),
+            ]
+        }
+        if values:
+            vals.update(values)
+        return cls.env['account.tax'].create(vals)
+
+    @classmethod
+    def make_downpayment(cls, **kwargs):
+        payment_params = {
+            'advance_payment_method': 'percentage',
+            'amount': 50,
+            **kwargs,
+        }
+        downpayment = cls.env['purchase.advance.payment.wizard'].with_context(cls.context).create(payment_params)
+        downpayment.create_invoices()
+        cls.purchase_order.button_confirm()
```

Comment:
It feels weird to me to directly reference the cls.record in a helper method like this. It's maybe better/safer to pass it as an argument every time to avoid unexpected behavior, e.g. someone calls the method and then tries to edit the record without realizing its changed due to the method. I'm not not sure if it may lead to unexpected behavior between rollbacks